File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), count(0)
    {
        subscriber = this->create_subscription<example_interfaces::msg::Int64>("number", 10,
                std::bind(&NumberCounterNode::callbackNumber, this, std::placeholders::_1));
        publisher = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        server = this->create_service<example_interfaces::srv::SetBool>(
                "reset_counter",
                std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
        RCLCPP_INFO(this->get_logger(), "Number counter has started");
    }

private:
    void callbackNumber(const example_interfaces::msg::Int64::SharedPtr numberMsg)
    {
        this->count += numberMsg->data;
        auto msg = example_interfaces::msg::Int64();
        msg.data = this->count;
        publisher->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Publishing count %ld", msg.data);
    }

    void callbackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
            const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data) {
            this->count = 0;
            response->message = "Reset count to zero";
        } else {
            response->message = "No change";
        }
        RCLCPP_INFO(this->get_logger(), "Reset counter: %s", response->message.c_str());
    }

    int count;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber;
    rclcpp::TimerBase::SharedPtr timer;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.count = 0;
        self.subscriber = self.create_subscription(
                Int64, "number", self.callback_number, 10)
        self.publisher = self.create_publisher(Int64, "number_count", 10)
        self.server = self.create_service(SetBool, "reset_counter",
                self.callback_reset_counter)
        self.get_logger().info("Number counter has started")

    def callback_number(self, number_msg):
        self.count += number_msg.data;
        count_msg = Int64()
        count_msg.data = self.count
        self.publisher.publish(count_msg)
        self.get_logger().info("Publishing count " + str(self.count))

    def callback_reset_counter(self, request, response):
        if request.data:
            self.count = 0
            response.message = "Reset count to zero"
        else:
            response.message = "No change"
        response.success = True
        self.get_logger().info(response.message)
        return response

def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

